Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+            case producing(
+                source: Producer.Source?,
+                pollClosure: (() -> Void)?,
+                running: Bool
+            )
+            /// The poll loop is currently suspended and we are waiting for an invocation
+            /// of `produceMore()` to continue producing messages.
+            case stopProducing(
+                source: Producer.Source?,
+                continuation: CheckedContinuation<Void, Error>?,
+                pollClosure: (() -> Void)?,
+                running: Bool
+            )
+            /// The system is shut down.
+            case finished
+        }
+
+        /// The current state of the state machine.
+        private var state: State = .uninitialized
+
+        /// Delayed initialization of `StateMachine` as the `source` and the `pollClosure` are
+        /// not yet available when the normal initialization occurs.
+        mutating func initialize(
+            source: Producer.Source,
+            pollClosure: @escaping () -> Void
+        ) {
+            guard case .uninitialized = self.state else {
+                fatalError("\(#function) can only be invoked in state .uninitialized, but was invoked in state \(self.state)")
+            }
+            self.state = .idle(source: source, pollClosure: pollClosure, running: false)
+        }
+
+        /// Actions to take after ``run()`` has been invoked on the ``KafkaPollingSystem/StateMachine``.
+        enum RunAction {
+            /// Do nothing, the poll loop is already terminated.
+            case alreadyClosed
+            /// The poll loop is already running. ``run()`` should not be invoked more than once!
+            case alreadyRunning
+            /// The poll loop can now be started.
+            case startLoop
+        }
+
+        mutating func run() -> RunAction {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state .uninitialized")
+            case .idle(let source, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .idle(source: source, pollClosure: pollClosure, running: true)
+            case .producing(let source, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .producing(source: source, pollClosure: pollClosure, running: true)
+            case .stopProducing(let source, let continuation, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .stopProducing(source: source, continuation: continuation, pollClosure: pollClosure, running: true)
+            case .finished:
+                return .alreadyClosed
+            }
+
+            return .startLoop
```

Comment:
Please add a return inside every individual case. That makes it easier to reason about and you don't have to scan until the end of the method